Document how the engineering team is organized (#798)
sourcegraph/about: handbook/engineering/organization.md

# Organization

The purpose of our engineering organization is to build a product that is valuable to our customers.

To achieve this purpose, we need to ensure two things:

> 1. Each feature or product that we build needs to coherently solve a particular customer problem. {#customer-needs}

Reviewer: is `{#customer-needs}` supposed to link somewhere? Same with `{#engineering-needs}` in the next line.

Author: The opposite: this creates an anchor on the page so that I can link to this part of the document from elsewhere.

Reviewer: 🤯 TIL

> 1. We make the necessary technology investments to maximize our velocity over time (e.g., building a standard component library, investing it automated testing). {#engineering-needs}

Reviewer: ```suggestion
1. We make the necessary technology investments to maximize our velocity over time (e.g., building a standard component library, investing in automated testing). {#engineering-needs}
```

> 1. We make the necessary technology investments to maximize our velocity over time (e.g., building a standard component library, investing in automated testing). {#engineering-needs}

Reviewer 2: ```suggestion
1. We make the necessary technology investments to maximize our velocity over time (e.g. building a standard component library, investing in automated testing). {#engineering-needs}
```

Author: I used to never put commas after e.g. and didn't think anything of it until I saw some other teammates (including @sqs) [doing it](https://sourcegraph.com/search?q=r%3Asourcegraph%2Fsourcegraph+e.g.%2C+count%3A1000&patternType=literal). Someone went so far as to [open a PR](https://github.com/sourcegraph/careers/pull/62/files) on our job descriptions to change it to have a comma. Since then I have just started adding a comma.

Based on an internet search, it seems like different style guides have different recommendations on this, so there isn't a single "right" way. If we want to be consistent as a company then we should add it to our style guide: https://about.sourcegraph.com/handbook/communication/style_guide

I started a poll in Slack to inform that decision: https://sourcegraph.slack.com/archives/C02FSM7DU/p1587578988157700

Going to leave this as-is on this PR.

This document describes how we organize engineering to ensure that we are making progress on both dimensions.

## Philosophy

This section describes how we think about our team in its ideal state. Due to a variety of real-world constraints that change over time (e.g., hiring, product priorities, engineering priorities, skills of our current team), our organization might temporarily exist in a state that is inconsistent with these ideals from time to time. That is ok.

### Teams

We have mission based teams. This means that each team is responsible for a particular long term mission, which may be a [product concern](#customer-needs) or an [engineering concern](#engineering-needs). In either case, the team contains the necessary cross-functional skillsets to achieve that mission. Each team is lead by a [Project Lead](roles.md#project-lead).

### Management

> [Engineering managers](roles.md#engineering-manager) are primarly responsible for the career development of the engineers who they manage. To best facilitate this goal, we prefer managers to focus on managing engineers of a particular skillset that the manager is familiar with. Manager are responsible for driving cross-team functional alignment to ensure that we are satisfying our [engineering needs](#engineering-needs) within the function that they manage.

Reviewer: ```suggestion
[Engineering managers](roles.md#engineering-manager) are primarily responsible for the career development of the engineers who they manage. To best facilitate this goal, we prefer managers to focus on managing engineers of a particular skill set that the manager is familiar with. Managers are responsible for driving cross-team functional alignment to ensure that we are satisfying our [engineering needs](#engineering-needs) within the function that they manage.
```
